### PR TITLE
reduce unwrap and create vector utils for handle special unwrap cases

### PR DESCRIPTION
**Pull request by giangndm-bluesea**
For better development in the feature, I think we should avoid using unwrap in logic code, which can cause server crashes (I think this is the main goal of str0m).

In this PR, I trying to move out some special unwrap cases (like push then get last_mut) to utils, which can be easier for controlling. Some code also can rewrite without unwrap version.

### CODE REVIEW

**Review by algesten — CHANGES_REQUESTED** [review]
Hi @giangndm-bluesea!

Thanks for taking the time to engage with the str0m code base like this!

Sadly I don't agree with the majority of these changes. I need to bring some of this reasoning out in the README.md, I will do that next. Str0m is built on a [fail fast](https://en.wikipedia.org/wiki/Fail-fast) principle.

That means when violating internal invariants and assumptions, rather than "limping on" in an unknown state, we panic. This makes it easier to find bugs, because the error can't be masked for some period of time and surface "in some other way".

From a library user's point of view, this should not matter since any incorrect user input should result in regular `RtcError`. The potential panics are strictly about internal state.

I realise this might not be clear for someone new to the code base, and I'm sorry you spent much time on it. I've tried to make a code comment for every unwrap, and I will update the README, but I'm interested to know if you think there are other things I could have done to make it clear this is desired behavior?

There are a couple of good changes in here, up to you if you want to amend the PR to only retain those, or maybe raise a new PR?

File: src/util/vector.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    fn push_last_get_mut(&mut self, item: T) -> &mut T;
+}
+
+impl<T> VecDequeExtended<T> for VecDeque<T> {
+    fn push_last_get_mut(&mut self, item: T) -> &mut T {
+        self.push_back(item);
+        self.back_mut()
+            .expect("should get last because of just pushed")
+    }
+}
```
[the remark this answers is not in the document]
Comment (k0nserv) [reply src/util/vector.rs:42]: I think `push_last_get_mut` is self explanatory, but it should be called `push_back_get_mut` to align with the existing `push_back` API. I do agree with @algesten's reasoning about short helpers, but this actually involves guarding against an invariant and repeating this code invites the possibility of incorrectness.

Given that it's only used in a single place I agree that the juice is not worth the squeeze here and we should get rid of this code

**Comment by giangndm-bluesea** [comment]
Thank @algesten  for a very careful review, I agree that we should fall fast when having any problems, but if we use the library for a huge media server that serves some thousand users then any single panic can make all of the users disconnected from the server (or maybe I misunderstood).

I think it can return Error instead of unwrap in some places for making the library more stable in any case.

**Comment by algesten** [comment]
> A broader point:
> 
> We seem to have quite a few cases of
> 
> ```rust
> // Comment explaining why unwrap is okay
> thing.unwrap();
> ```
> 
> I think changing these to
> 
> ```rust
> thing.expect("<reasoning>");
> ```
> 
> would be an improvement

Agree. I think there are three cases here:

1. The reason for `unwrap()` happened literally on the row preceding. An expect message might be just noise.
2. The reason for `.unwrap()` is quite complicated needing more explanation that can be condensed into a catchy "due to blah". I don't think it's helpful to write essays inside the .expect() comment.
3. The reason is easily condensed.

I think I should make a pass over the code and find 3 and maybe some 2.

**Comment by algesten** [comment]
> Thank @algesten for a very careful review, I agree that we should fall fast when having any problems, but if we use the library for a huge media server that serves some thousand users then any single panic can make all of the users disconnected from the server (or maybe I misunderstood).
> 
> I think it can return Error instead of unwrap in some places for making the library more stable in any case.

I agree. Some of these unwraps are definitely guarded by Error already, but I probably haven't commented in code enough to bring that out.

Panic in Rust is on _thread level_, which means it doesn't need to take the server down, just the thread. Granted, it won't necessarily free memory, locks etc as needed.

**Comment by k0nserv** [comment]
> Panic in Rust is on thread level, which means it doesn't need to take the server down, just the thread. _Granted, it won't necessarily free memory, locks etc as needed._

With panic unwinding it should.

File: src/ice/agent.rs [hunk 2]
```diff
@@ -703,7 +703,9 @@ impl IceAgent {
         // the ICE agent sending the request, separated by a colon (":").
         if message.is_binding_request() {
             // The existence of USERNAME is checked in the STUN parser.
-            let (local, remote) = message.split_username().unwrap();
+            let Some((local, remote)) = message.split_username() else {
+                return false;
+            };
```
[the remark this answers is not in the document]
Comment (giangndm-bluesea) [reply src/ice/agent.rs:708]: I agree that in this case message.split_username() should not return Option.

[the remark this answers is not in the document]
Comment (algesten) [reply src/ice/agent.rs:708]: Yes. We can change that.

**Comment by giangndm-bluesea** [comment]
@algesten sorry for late response, If the macro can provide more useful info so I think this PR will be canceled. If I found other, I will send another PR.

**Pull request closed without merging**